src/libcollections/btree/map.rs

```diff
@@ -529,6 +537,174 @@ impl<K: Ord, V> BTreeMap<K, V> {
+                                            cur_node = parent.forget_type();
+                                        }
+                                    },
+                                    Err(node) => {
+                                        // We are at the top, create a new root node and subtree.
+                                        let tree_height = node.height();
+                                        let mut new_root = node.into_root_mut().push_level();
+
+                                        let mut right_tree = node::Root::new_leaf();
+                                        for _ in 0..tree_height {
+                                            right_tree.push_level();
+                                        }
+
+                                        // Push key-value pair and new subtree.
+                                        new_root.push(key, value, right_tree);
+                                        cur_node = new_root.last_edge().descend();
+                                        break;
+                                    },
+                                }
+                            }
+
+                            // Go down to the right-most leaf again.
+                            loop {
+                                match cur_node.force() {
+                                    Leaf(leaf) => {
+                                        cur_node = leaf.forget_type();
+                                        break;
+                                    },
+                                    Internal(internal) => cur_node = internal.last_edge().descend(),
+                                }
+                            }
+                        }
+                    },
+                    Internal(_) => unreachable!(),
+                }
+
+                length += 1;
+            }
+        }
+        {
+            // Handle underfull nodes, start from the top.
+            let mut cur_node = root.as_mut();
+            loop {
+                match cur_node.force() {
+                    Leaf(_) => break,
+                    Internal(internal) => {
+                        // Check if right-most child is underfull.
+                        let right_child = internal.last_edge().descend();
+                        if right_child.len() < node::CAPACITY / 2 {
+                            // We need to steal.
+                            let num_steals = node::CAPACITY/2 - right_child.len();
+                            match right_child.ascend() {
+                                Ok(parent) => {
+                                    let mut parent = match parent.left_kv() {
+                                        Ok(left) => left,
+                                        Err(_) => unreachable!(),
+                                    };
+
+                                    for _ in 0..num_steals {
+                                        parent = parent.steal_left();
```

Comment:
This loop is not the most efficient way to do bulk stealing, and is in fact quadratic in the number of elements stolen. Thankfully, since `CAPACITY` is constant, this isn't really any sort of asmyptotic problem, but it might be worth looking into the creation of some method that would do many steals at once.